14 add tests

apast opened the pull request:
Hi,

I added initial tests over command.py (coverage 98%) and to __main__.py.

Tests are implemented using basic unittest, and I run it on Linux using python 2 & 3.

I am implementing it from top-down, before try to add or change any funcionality.

General structure of code was keep as the same. I only fixed docopt [default: ...] definitions for --spool and --ext and now we have default values defined by documentation and being set by docopt parser. I updated README.md accordingly.

I mocked pytest_watch.watcher.watch and pytest_watch.config.merge_config to isolate and focus on CLI input and validadion.

Can you review and give me some feedback to improve the code before accept this PR?

Thank you!

joeyespo commented on pytest_watch/command.py:
```diff
@@ -29,7 +29,7 @@
   --spool <delay>       Re-run after a delay (in milliseconds), allowing for
-                        more file system events to queue up (default: 200 ms).
+                        more file system events to queue up [default: 200].
   -p --poll             Use polling instead of OS events (useful in VMs).
```
Ah, nice removal of the redundant `ms`.

joeyespo commented on pytest_watch/command.py:
```diff
@@ -11,7 +11,7 @@
                         collection (multi-allowed).
   --ext <exts>          Comma-separated list of file extensions that can
-                        trigger a new test run when changed (default: .py).
+                        trigger a new test run when changed [default: .py].
```
~Are brackets a common for default values in CLI tool `--help` texts?~ Nevermind. Just noticed docopt uses brackets [in their docs](http://docopt.org/). Thanks for updating that.

joeyespo commented on pytest_watch/__main__.py:
```diff
@@ -8,12 +8,14 @@
 
     from pytest_watch.command import main
-    main()
+    main(argv=sys.argv[1:])
```
I think `main()` with no args is more aesthetic. The default is to use `argv[1:]` and adding that here adds noise.

Could you change it back?

If your rationale was "I couldn't tell what main's default were", a better docstring for `main` could address that instead of more code.

joeyespo commented on pytest_watch/command.py:
```diff
@@ -87,23 +91,24 @@ def main(argv=None):
         extensions = [('.' if not e.startswith('.') else '') + e
                       for e in args['--ext'].split(',')]
-    else:
-        extensions = None
```
Ahh I see fixing the default value in docopt allows you to remove this. Very nice 👍

joeyespo commented on pytest_watch/command.py:
```diff
@@ -87,23 +91,24 @@ def main(argv=None):
+    try:
+        spool = int(spool)
+    except ValueError:
+        sys.stderr.write('Error: Spool (--spool %s) must be an integer.\n' % spool)
```
Two things.

1. I used to willfully break PEP8's "80 character limit" rule all the time. With side-by-side comparisons and 3-way merges, I've since come to appreciate this style rule. Could you do the same here?

2. I thought I read a while back that Python favored `.format` over `%` these days. A quick Google search yields no results on this. But since I've since switched to using `.format` everywhere, could you do the same here for consistency? A case could be made later to go back to `%` everywhere.

This should do it:

```
        sys.stderr.write(
            'Error: Spool (--spool {}) must be an integer.\n'.format(spool))
```

Or perhaps:

```
        print(
            'Error: Spool (--spool {}) must be an integer.\n'.format(spool),
            file=sys.stderr)
```

joeyespo commented on pytest_watch/command.py:
```diff
@@ -87,23 +91,24 @@ def main(argv=None):
+
+    if spool < 0:
+        sys.stderr.write("Error: Spool value(--spool %s) must be positive integer" % spool)
+        return 2
```
Ahh nice catch! 👍 

Also, same comment as above about using 80 characters.

joeyespo commented on pytest_watch/tests/test_main.py:
```diff
@@ -0,0 +1,20 @@
+from pytest_watch.__main__ import run_cli
+
+
+class TestRunCLI(unittest.TestCase):
```
Is there any chance you could switch to classless pytest? Using classes just adds overhead IMHO.

If not, I'll probably go back and do this at a (much) later date.

apast replied:
Yes, I will adopt this classless where it is unecessary.

joeyespo commented on pytest_watch/watcher.py:
```diff
@@ -184,10 +184,16 @@ def run_hook(cmd, *args):
+    ignore = [] if ignore is None else ignore
+    extensions = [] if extensions is None else extensions
+    pytest_args = [] if pytest_args is None else pytest_args
+
```
This is much better than what I had 👍 Avoiding any accidental mutation is definitely a good thing.

joeyespo commented on pytest_watch/tests/test_command.py:
```diff
@@ -0,0 +1,437 @@
+import sys
```
(Same comment in this file about using functions over class-based tests.)

Also `assert x != y` is much more readable than `self.assertFalse(condition)`, etc. Pytest does an amazing job of inspecting values when things go wrong, so I'm all for the boost in readability using raw `assert`.

Again, if you'd rather get this in now, that's fine. I'll probably go back and switch it later on if that's the case.

apast replied:
Perfect. I agree with readability os simple assert. I will adopt this for general assertions.

I am in doubt about self.assertListEqual and self.assertDictEqual calls. I think it is better to use unittest helpers in this case. What do you think?

Thanks

joeyespo replied:
Well given that [pytest advocates the use of `assert`](https://docs.pytest.org/en/latest/assert.html) over `self.assert`, and this is a pytest-based project, I think using vanilla `assert` is appropriate. Here's the quote directly from the [Features list](https://docs.pytest.org/en/latest/#id1): "Detailed info on failing assert statements (no need to remember `self.assert*` names)"

FWIW readability was one of the things that initially attracted me to pytest to begin with.

apast replied:
Done!

joeyespo commented on pytest_watch/command.py:
```diff
@@ -87,23 +91,24 @@ def main(argv=None):
         pytest_args.append('--pdb')
 
-    # Parse extensions
+    # Parse extensions [default: .py]
```
Nice addition of comments 👍 Even more so with the defaults right there.

joeyespo commented on .gitignore:
```diff
@@ -14,3 +14,7 @@ env
 Thumbs.db
+.coverage
+.coverage.*
+venv
```
"Environment file" may be a better place for `venv`.

I wonder if it'd make sense to add a new "Testing files" for things like `.coverage`.

apast replied:
Can we adopt this https://github.com/github/gitignore/blob/master/Python.gitignore file in pytest_watch project?

joeyespo commented:
Nice! Just tested on Windows and everything is passing 💯 😌 

The only thing I had to do was install `pytest-cov`. I think we can signal that requirement by:
- Adding a new `requirements-test.txt` file
- Adding `pytest-cov>=2.5.1` to it
- Adding a [tests_require field](https://setuptools.readthedocs.io/en/latest/setuptools.html?highlight=tests_require#new-and-changed-setup-keywords) to `setup.py`:
  ```
  tests_require=read('requirements-test.txt').splitlines()
  ```
  (Note that after doing this, running `python setup.py test` will succeed even if you don't manually install pytest-cov for pytest to use.)

joeyespo commented:
@blueyed Anything you'd like to add?

apast commented:
Hello, Joe,

Thanks for your detailed review.

I will check this and change code or reply properly.

[]'s,

Andre

apast commented:
Joe,

it seems excellent to add this requirements-test.txt file and declare it in setup.py file.

I will implement this suggestion and commit soon.

[]'s,
Andre

apast commented:
I added a version checker inside setup.py for python 2 compatibility.

I really don't know if it is the best pratice, or can we keep it this way for this moment. What do you think?

[]'s,

joeyespo commented on pytest_watch/__main__.py:
```diff
@@ -8,12 +8,14 @@
     from pytest_watch.command import main
     main()
+
+if __name__ == '__main__': run_cli()
```
Oh, meant to say before. Could you move `run_cli` to its own line?

I generally find that more readable. [PEP8 also prefers it](https://www.python.org/dev/peps/pep-0008/#multiline-if-statements). (Search for "Compound statements (multiple statements on the same line) are generally discouraged".)

apast replied:
done!

apast commented:
@joeyespo , I am working on new tests over watcher.py file. It would be great to have a feedback (or approval) about these tests style to replicate over new features and tests I am implementing.

I am think to extract the helpful guidelines you wrote over this PR and start a project guideline. i.e.: builtin assert, classless, ...

apast commented:
#14 Add tests. Target code: watcher.py
This test coverage started at [commit](https://github.com/joeyespo/pytest-watch/pull/79/commits/93614122dd2be0ed8e3e27956adfd720d8ad793f)

The main purpose of this PR extension is coverage of file pytest_watch/watcher.py, aiming to achieve #14 issue purposes.

I achieved 76% of coverage at this moment.

It is pending the implementation of tests over queue.empty() and dequeue_all() calls, and observers. It is in WIP state.

Please, give me feedback and followup and if is what is necessary to get this test codebase into master branch, improving project guidelines and test coverage for new contributions.

I installed travis-ci support over my own repository to check status of branchs. It is being tested over python 3.6 and 2.7 versions. Build status can be checked here.

My best,

Andre

apast commented on pytest_watch/watcher.py:
```diff
@@ -117,10 +119,18 @@ def _reduce_events(events):
+
 def _show_summary(argv, events, verbose=False):
     command = ' '.join(argv)
-    bright = lambda arg: STYLE_BRIGHT + arg + Style.RESET_ALL
```
__________ PEP8-check __________
pytest_watch/watcher.py:124:5: E731 do not assign a lambda expression, use a def
bright = lambda arg: STYLE_BRIGHT + arg + Style.RESET_ALL
^

apast replied:
just for pep8 purposes

apast commented on pytest_watch/watcher.py:
```diff
@@ -117,10 +119,18 @@ def _reduce_events(events):
 def _show_summary(argv, events, verbose=False):
     command = ' '.join(argv)
-    bright = lambda arg: STYLE_BRIGHT + arg + Style.RESET_ALL
-    highlight = lambda arg: STYLE_HIGHLIGHT + arg + Style.RESET_ALL
```
__________ PEP8-check __________
pytest_watch/watcher.py:125:5: E731 do not assign a lambda expression, use a def
highlight = lambda arg: STYLE_HIGHLIGHT + arg + Style.RESET_ALL
^

apast replied:
just for pep8 purposes

apast commented:
Hello @joeyespo ,

there are more changes to implement over this PR? I wanna follow in more tests handle Issues, but, before any change on core code, I need to know if we can proceed with this test base.

Can you check this PR?
Thanks,
Andre

blueyed commented on setup.py:
```diff
@@ -19,6 +29,7 @@ def read(filename):
     platforms='any',
     packages=find_packages(),
     install_requires=read('requirements.txt').splitlines(),
+    tests_require=get_test_requirements(),
```
I would add the requirements in setup.py, both for install and tests.
As for tests, there could then also be an `extras_require`, e.g.:
```python
    extras_require={
        'testing': DEPS_TESTING,
        'dev': DEPS_TESTING + DEPS_QA + [
            'pdbpp',
            'pytest-pdb',
        ],
        'qa': DEPS_QA,
    },
```

This allows to `pip install -e '.[testing]'` then.
`pip install -e .` uses the existing requirements (which should be moved from requirements.txt) to setup.py.

blueyed commented:
As for the tests: why is it using unittest that much, and not only pytest?

blueyed commented on .gitignore:
```diff
@@ -1,16 +1,106 @@
-env
-.cache
-*.so
+# source: https://github.com/github/gitignore/blob/master/Python.gitignore
```
I find this too verbose/much.
Most of it makes sense in a user's global gitignore already/only.

apast commented:
Really! There is a lot of  unecessary unittest and unittest.TesCase inheritance declared.

Initially, I had choose unittest model for portability for both frameworks. But, as Joe wrote at some code review, it is a pytest project, and it would be good if it follows it's principles.

I agree with you.

I removed some statements and important for unittest root module, renamed setup and teardown methods and TestCase inheritance.

For this moment, what do you think if we keep unittest.mock usage and move to pytest.fixtures for further project tests coverage evolution?

Thanks for your help,
Andre

blueyed commented:
mock is fine in general, but you could use https://github.com/pytest-dev/pytest-mock on top.. :)

apast commented:
Now, using pytest-mock fixture!

apast commented:
@blueyed @joeyespo I added some more tests. Now, we have almost 90% of coverage.

Coverage evolution can be checked here: https://codecov.io/gh/apast/pytest-watch/branch/14-add-tests
Build status, here: https://travis-ci.org/apast/pytest-watch

I didn't test it over Windows. Please, if you can, please, test and give me a followup for python versions 2.7 and 3.6.

For any further consideration before merge this PR, please, let me know for we advance on issues resolution.

apast commented:
There are three inconsistencies on tests on Windows.

I am working on it.

Two tests are related to same_path check for symlinks. It was marked with skip for Windows environment.

Now, I am working on sys.executable path issue when it contains whitespaces, like c:\Program Files.

Hope to share it soon.

apast commented:
It is ok on Windows, again.

@blueyed @joeyespo Please, can you check this and give some feedback?

Thanks,
Andre

apast commented:
I just added initial Windows test automation with AppVeyor. I believe that it should run by tox runner for multiple python version support.
For this moment, I can only contribute with this. Hope to contribute more soon.

It would be nice if someone review current version before any code improvement and fixes.

joeyespo commented:
Sorry, lots to review! Getting through it slowly. I do appreciate the work put into this.

apast commented:
I agree. It is a long PR.

You helped during this evolution.

Thanks!
